[c++-interop] For failed imports in ClangImporter, cache them regardless.

plotfi opened the pull request:
As per SR-14137 this caches entries in ImportedDecls even when the
import failed. The use of Optional<NullablePtr<Decl>> is employed here
as a better way to track this than a raw nullptr.

Also have to mention I did this based on Thomas's PR #36747.

This in theory should help us better handle complex templates and
dependant types.

plotfi commented:
@zoecarver @hyp @CodaFi

plotfi commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -8361,7 +8372,8 @@ Decl *ClangImporter::Implementation::importDeclCached(
   if (Known != ImportedDecls.end())
-    return Known->second;
+    if (Known->second.hasValue())
+    return Known->second.getValue().get();
```
nit: indentation

plotfi commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -9236,8 +9248,10 @@ Decl *ClangImporter::Implementation::importDeclAndCacheImpl(
+  if (!Result) {
+    ImportedDecls[{Canon, version}] = None;
     return nullptr;
+  }
```
This is the cache the failed decls. Is there anything else that should be cached here or anywhere else other than the Canon and version?

zoecarver commented:
Storing a `Optional<NullablePtr>` in the map of imported decls doesn't make sense. That already has two failures modes: 1) does the key exists? 2) Is the value null? 

What you want is to return an `Optional<NullablePtr>` from `importDeclCached`. That way you can communicate both: 1) is there a cached thing here and 2) is that thing null?

I think you can just take Thomas' patch as-is and try to land it. IIRC there were some test failures, so the work in landing this is just debugging the test failures and landing it.

plotfi committed:
[c++-interop] For failed imports in ClangImporter, cache them regardless.

As per SR-14137 this caches entries in ImportedDecls even when the
import failed.

Also have to mention I did this based on Thomas's PR #36747.

This should help us better handle complex templates and dependant types.

plotfi commented:
> What you want is to return an `Optional<NullablePtr>` from `importDeclCached`. That way you can communicate both: 1) is there a cached thing here and 2) is that thing null?

Updated patch, but no optionals at all. I can update importDeclCached to provide a tri-state through Optional<Decl *> where None means nothing was found and nullptr/non-nullptr means it either found a cached import-failure or a succeeded imported decl. IIRC Thomas' patch did not do this (it just returns None if there was a cached nullptr), which is why I think this confused me. I'll ping this PR once I rerun the test suite on this newest version.

plotfi commented:
> Storing a `Optional<NullablePtr>` in the map of imported decls doesn't make sense. That already has two failures modes: 1) does the key exists? 2) Is the value null?
> 
> What you want is to return an `Optional<NullablePtr>` from `importDeclCached`. That way you can communicate both: 1) is there a cached thing here and 2) is that thing null?
> 
> I think you can just take Thomas' patch as-is and try to land it. IIRC there were some test failures, so the work in landing this is just debugging the test failures and landing it.

Updated to provide the additional information about prior import failures from `importDeclCached`. I ran a ninja check-swift and I don't see any additional test failures.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4624,8 +4625,12 @@ namespace {
       // itself.
-      if (auto Known = Impl.importDeclCached(decl, getVersion()))
-        return Known;
+      auto Known = Impl.importDeclCached(decl, getVersion());
+      if (Known.hasValue() && Known.getValue()) {
+        return Known.getValue();
+      } else if (Known.hasValue()) {
+        // TODO: Handle case where we have a failed-cached-import here.
+      }
```
```suggestion
      auto Known = Impl.importDeclCached(decl, getVersion());
      if (Known.hasValue())
        return Known.getValue();
```

plotfi replied:
Ah ok. I can do this. I was thinking of leaving the TODO blocks to hook up to the diag in a later patch; sorry should have been more clear here.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -5667,8 +5674,12 @@ namespace {
       // itself.
-      if (auto Known = Impl.importDeclCached(decl, getVersion()))
-        return Known;
+      auto Known = Impl.importDeclCached(decl, getVersion());
```
same

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -5792,7 +5803,8 @@ namespace {
         if (known != Impl.ImportedDecls.end())
-          return known->second;
+          if (known->second)
+            return known->second;
```
Same: if we have a cached decl, return it (even if it's a nullptr).

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -8354,16 +8366,16 @@ void SwiftDeclConverter::importInheritedConstructors(
+    return None;
 
-  return nullptr;
+  return Known->second ? Known->second : nullptr;
```
I don't think we need the ternary. Just `return Known->second`.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -9223,11 +9235,14 @@ Decl *ClangImporter::Implementation::importDeclAndCacheImpl(
       return nullptr;
-    return Known;
+    return Known.getValue();
+  } else if (Known.hasValue()) {
```
same

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -9405,8 +9422,12 @@ ClangImporter::Implementation::importDeclForDeclContext(
+  auto cached = importDeclCached(contextDecl, version, useCanonicalDecl);
+  if (cached.hasValue() && cached.getValue()) {
+    return cached.getValue();
+  } else if (cached.hasValue()) {
```
same

plotfi replied:
`ClangImporter/attr-swift_name.swift` fails if we return null from here because the diagnostics code for swift_name_circular_context_import doesn't get hit to append the warning+notes:

```
/Users/plotfi/local/S/swift/test/ClangImporter/Inputs/custom-modules/ObjCIRExtras.h:78:52: warning: cycle detected while resolving 'MutuallyCircularNameA' in swift_name attribute for 'MutuallyCircularNameB'
SWIFT_NAME(MutuallyCircularNameA.Inner) @interface MutuallyCircularNameB : NSObject @end

/Users/plotfi/local/S/swift/test/ClangImporter/Inputs/custom-modules/ObjCIRExtras.h:77:52: note: while resolving 'MutuallyCircularNameB' in swift_name attribute for 'MutuallyCircularNameA'
SWIFT_NAME(MutuallyCircularNameB.Inner) @interface MutuallyCircularNameA : NSObject @end

/Users/plotfi/local/S/swift/test/ClangImporter/Inputs/custom-modules/ObjCIRExtras.h:78:52: note: please report this issue to the owners of 'ObjCIRExtras'
SWIFT_NAME(MutuallyCircularNameA.Inner) @interface MutuallyCircularNameB : NSObject @end
```

Since this function will return nullptr from this point regardless, do we want to bail before emitting these warnings?

zoecarver commented:
> Updated patch, but no optionals at all. I can update importDeclCached to provide a tri-state through Optional<Decl *> where None means nothing was found and nullptr/non-nullptr means it either found a cached import-failure or a succeeded imported decl. IIRC Thomas' patch did not do this (it just returns None if there was a cached nullptr), which is why I think this confused me. I'll ping this PR once I rerun the test suite on this newest version.

Yeah makes sense. In this case, you can return `Optional.some(nullptr)` when we've cached a nullptr. Make sense?

zoecarver commented:
This looks good, though. Thanks again, Puyan!

plotfi commented:
> Yeah makes sense. In this case, you can return `Optional.some(nullptr)` when we've cached a nullptr. Make sense?

Do we want to actually return nullptr in the hasValue() == true getValue() == nullptr cases for all of the following:

```
Decl *VisitObjCMethodDecl(const clang::ObjCMethodDecl *decl)
Decl *VisitObjCPropertyDecl(const clang::ObjCPropertyDecl *decl)
Decl *ClangImporter::Implementation::importDeclAndCacheImpl()                                                                                                                                                                                         ClangImporter::Implementation::importDeclForDeclContext()
```

In that case there are lit failures. If we catch the nullptrs and continue as before then there are no failures. 

I think probably for the top level Visitors we may want to predicate this on `if (hasValue() && getValue())` if we return getValue().

plotfi commented:
@zoecarver How does this look? Do you approve of the test case modification? From debugging the flow of the importer with the nullptr caching I believe the change in number of swift_name_circular_context_import warnings is due to the caching.

plotfi commented:
@zoecarver Do you think this patch needs more work? I've done some investigation on the swift_name_circular_context_import warnings and I am not sure the exact same ordering of warnings can be printed out with the added caching.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3417,7 +3417,8 @@ namespace {
       if (alreadyImportedResult != Impl.ImportedDecls.end())
-        return alreadyImportedResult->second;
+        if (alreadyImportedResult->second)
+          return alreadyImportedResult->second;
```
We can't return nullptr here?

zoecarver replied:
Maybe we can re-factor this to use `Impl.importDeclCached`?

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4830,8 +4832,9 @@ namespace {
                                               getVersion()});
         if (known != Impl.ImportedDecls.end()) {
```
If we already have a known nullptr, maybe it should just return nullptr?

plotfi replied:
This one actually changes the behavior pretty significantly. Lot of tests fail. Do we want it to stop here and return nullptr instead of continuing with all the code that happens after `ImportedName importedName;` or 
`// Determine whether the function is throwing and/or async.` (for the next one you commented on) ?

plotfi replied:
> Oh also, the logic of this code was to return the decl if `isAcceptableResult` returns true. If the decl is nullptr due to a bad import attempt can we exactly consider this nullptr as a meeting these criteria:
> 
> ```
>     static bool isAcceptableResult(Decl *fn,
>                                    Optional<AccessorInfo> accessorInfo) {
>       // We can't safely re-use the same declaration if it disagrees
>       // in accessor-ness.
>       auto accessor = dyn_cast<AccessorDecl>(fn);
>       if (!accessorInfo)
>         return accessor == nullptr;
> 
>       // For consistency with previous behavior, allow it even if it's been
>       // imported for some other property.
>       return (accessor && accessor->getAccessorKind() == accessorInfo->Kind);
>     }
> ```
> 
> ?

Hmm, on second thought I think this could be the case given:

```
      auto accessor = dyn_cast<AccessorDecl>(fn);
      if (!accessorInfo)
        return accessor == nullptr;

```

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4982,8 +4985,9 @@ namespace {
           auto decl = known->second;
-          if (isAcceptableResult(decl, accessorInfo))
-            return decl;
+          if (nullptr != decl)
```
Same as above.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -5800,7 +5805,7 @@ namespace {
         auto known = Impl.ImportedDecls.find({decl->getCanonicalDecl(),
                                               getVersion()});
-        if (known != Impl.ImportedDecls.end())
+        if (known != Impl.ImportedDecls.end() && known->second != nullptr)
```
Same as above. Why can't we return nullptr here?

plotfi replied:
I actually only added this one back because Thomas had it in the previous patch. Everything passes fine with this line returning nullptr.

zoecarver commented:
> Do you approve of the test case modification?

I'll take a closer look, but it seems fine to me.

zoecarver commented:
@swift-ci please test.

zoecarver commented:
By the way this patch looks great. Seems like it doesn't touch most visitors which is awesome!

Thanks again for working on it.

plotfi commented:
> By the way this patch looks great. Seems like it doesn't touch most visitors which is awesome!
> 
> Thanks again for working on it.

@zoecarver I addressed your comments. For some of them I just altered isAcceptableResult to do the thing that makes most sense when a nullptr entry is found (if its not an AccessorDecl return false).

hyp commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4761,7 +4762,7 @@ namespace {
       // We can't safely re-use the same declaration if it disagrees
       // in accessor-ness.
-      auto accessor = dyn_cast<AccessorDecl>(fn);
+      auto accessor = dyn_cast_or_null<AccessorDecl>(fn);
```
I had a hard time understanding the new semantics of `isAccceptableResult` with this patch. It looks like it will return `true` when fn is null, which is what we want, but it's not very clear from the code. What if we rename this function to `isAcceptableResultOrNull` and adding an early return for `fn == nullptr` instead of using `dyn_cast_or_null`?

plotfi replied:
Cool yeah. That'd be less of a subtile meaning here.

drodriguez commented:
@swift-ci please smoke test

drodriguez commented:
@swift-ci please test Linux platform

plotfi commented:
@hyp @zoecarver Tests passing, ready to merge unless you all have more review comments.

hyp reviewed (approved):
LGTM, thanks
